File: third/thirdsync/src/main/java/joe/sync/mapper/ThirdMapper.java
package joe.sync.mapper;

import joe.reactive.dto.ThirdDto;
import joe.reactive.fourth.FourthDto;
import org.mapstruct.Mapper;

import java.time.LocalDateTime;

@Mapper(componentModel = "spring")
public interface ThirdMapper {
    default FourthDto map(ThirdDto thirdDto, String callbackUrl) {
        return FourthDto.builder()
                .field1111(thirdDto.getField111())
                .field2222(thirdDto.getField222())
                .field3333(thirdDto.getField333())
                .field4444(thirdDto.getField444())
                .start(LocalDateTime.now())
                .callbackUrl(thirdDto.isNeedCallback()?callbackUrl:null)
                .build();
    }
}

File: third/thirdredis/src/main/java/redis/mapper/ThirdMapper.kt
package redis.mapper

import joe.reactive.dto.ThirdDto
import joe.reactive.fourth.FourthDto
import org.springframework.stereotype.Component

@Component
class ThirdMapper {
    fun map(thirdDto: ThirdDto): FourthDto {
        return FourthDto.builder()
            .field1111(thirdDto.field111)
            .field2222(thirdDto.field222)
            .field3333(thirdDto.field333)
            .field4444(thirdDto.field444)
            .build()
    }
}